yarn-project/noir-protocol-circuits/src/crates/private-kernel-lib/src/private_kernel_inner.nr

```diff
@@ -26,22 +26,23 @@ impl PrivateKernelInputsInner {
     }
 
     fn validate_contract_tree_root(self) {
-        let purported_contract_tree_root = self.private_call.call_stack_item.public_inputs().historical_block_data.contract_tree_root();
+        let purported_contract_tree_root = self.private_call.call_stack_item.public_inputs.historical_block_data.contract_tree_root();
         let previous_kernel_contract_tree_root = self.previous_kernel.public_inputs.constants.block_data.contract_tree_root();
 
         assert(purported_contract_tree_root == previous_kernel_contract_tree_root, "purported_contract_tree_root does not match previous_kernel_contract_tree_root");
     }
 
     fn validate_inputs(self) {
         let this_call_stack_item = self.private_call.call_stack_item;
-        assert(this_call_stack_item.function_data().is_private, "Cannot execute a non-private function with the private kernel circuit");
+        let function_data = this_call_stack_item.function_data;
+        assert(function_data.is_private, "Private kernel circuit can only execute a private function");
+        assert(function_data.is_constructor == false, "A constructor must be executed as the first tx in the recursion");
+        assert(self.previous_kernel.public_inputs.is_private, "Can only verify a private kernel snark in the private kernel circuit");
 
-        let start = self.previous_kernel.public_inputs.end;
-        let start_private_call_stack_length = array_length(start.private_call_stack);
-
-        assert(self.previous_kernel.public_inputs.is_private, "Cannot verify a non-private kernel snark in the private kernel circuit");
-        assert(this_call_stack_item.function_data().is_constructor == false, "A constructor must be executed as the first tx in the recursion");
-        assert(start_private_call_stack_length != 0, "Cannot execute private kernel circuit with an empty private call stack");
+        // An empty stack will fail at self.pop_and_validate_this_private_call_hash.
+        // let start = self.previous_kernel.public_inputs.end;
+        // let start_private_call_stack_length = array_length(start.private_call_stack);
+        // assert(start_private_call_stack_length != 0, "Cannot execute private kernel circuit with an empty private call stack");
```

Comment:
Yeah the pop should check the length automatically so we should be able to remove this :D